### PR TITLE
New RA for mpathpersist

### PR DESCRIPTION
**Pull request by ls-zhu**
It is a new RA hanlding persist reservation for multipath
devices. It uses mpathpersist which from package multipath-tools
as backend.
fate#324044

### CODE REVIEW

**Comment by krig** [comment]
Hi Lingshan, please add the new agent to the makefiles:

`heartbeat/Makefile.am`
`doc/man/Makefile.am`

File: heartbeat/mpathpersist [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+
+# set default values
+    : ${mpathpersist_binary="mpathpersist"} # binary name for the resource
+    : ${devs=""}                            # device list
+    : ${required_devs_nof=1}                # number of required devices
+    : ${reservation_type=1}                 # reservation type
+    : ${master_score_base=0}                # master score base 
+    : ${master_score_dev_factor=100}        # device factor for master score
+    : ${master_score_delay=30}              # delay for master score
```
Comment (oalbrigt) [inline heartbeat/mpathpersist]: You should create _default variables to assign from here (and also reuse in metadata, so you only have to change it once in future updates). See https://github.com/ClusterLabs/resource-agents/blob/master/heartbeat/nagios#L20 for an example.

File: heartbeat/mpathpersist [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+#######################################################################
+# Initialization:
+
+: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+
+# set default values
+    : ${mpathpersist_binary="mpathpersist"} # binary name for the resource
+    : ${devs=""}                            # device list
+    : ${required_devs_nof=1}                # number of required devices
```
Comment (oalbrigt) [inline heartbeat/mpathpersist]: It seems like required_devs_nof is a typo (required_devs_no and required_devs_nof are used in this agent, but only nof has a default).

File: heartbeat/mpathpersist [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+#    OCF_RESKEY_required_devs_no
+#    OCF_RESKEY_reservation_type
+#    OCF_RESKEY_master_score_base
+#    OCF_RESKEY_master_score_dev_factor
+#    OCF_RESKEY_master_score_delay
+#
+# TODO
+# 
+# 1) PROBLEM: devices which were not accessible during 'start' action, will be never registered/reserved 
+#    TODO:    'Master' and 'Salve' registers new devs in 'monitor' action
```
Comment (krig) [inline heartbeat/mpathpersist]: Typo: 'Salve' -> 'Slave'

File: heartbeat/mpathpersist [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+
+<longdesc lang="en">
+This resource agent manages SCSI PERSISTENT RESERVATIONS ON MULTIPATH DEVICES.
+"mpathpersist" from multipath-tools is used, please see its documentation.
+Should be used as multistate (Master/Slave) resource
+Slave registers its node id ("crm_node -i") as reservation key ( --param-sark ) on each device in the params "devs" list.
+Master reservs all devices from params "devs" list with reservation "--prout-type" value from "reservation_type" parameter.
+Please see man SG_PERSIST(8) and MPATHPERSIST(8) for reservation_type details.
+</longdesc>
+<shortdesc lang="en">Manages SCSI PERSISTENT RESERVATIONS ON MULTIPATH DEVICES</shortdesc>
```
Comment (krig) [inline heartbeat/mpathpersist]: Why all caps here? Please lower case this sentence :)

File: heartbeat/mpathpersist (hunk 2 above)
Comment (ls-zhu) [reply heartbeat/mpathpersist]: fixed, thanks

File: heartbeat/mpathpersist [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+Minimum number of "working" devices from device list
+      1) existing 
+      2) "mpathpersist --in --read-keys \$device" works (Return code 0)
+resource actions "start","monitor","promote" and "validate-all" return "\$OCF_ERR_INSTALLED"
+if the actual number of "working" devices is less then "required_devs_no".
+resource actions "stop" and "demote" tries to remove reservations and registration keys from 
+all working devices, but always return "\$OCF_SUCCESS"
+</longdesc>
+<shortdesc lang="en">minimum number of working devices</shortdesc>
+<content type="string" default="1"/>
```
Comment (krig) [inline heartbeat/mpathpersist]: Should also use `OCF_RESKEY_required_devs_no_default` here (same for other parameters)

File: heartbeat/mpathpersist [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+Please see man sg_persist(8) and mpathpersist(8) for reservation_type details.
+</longdesc>
+<shortdesc lang="en">Manages SCSI persistent reservations on multipath devices</shortdesc>
+
+<parameters>
+<parameter name="binary" unique="0">
+<longdesc lang="en">
+The name of the binary that manages the resource.
+</longdesc>
+<shortdesc>the binary name of the resource</shortdesc>
```
Comment (krig) [inline heartbeat/mpathpersist]: The `<shortdesc>` tag requires the `lang="en"` attribute to validate.

Example fix: https://github.com/krig/resource-agents/commit/3b0ccb011def18842f64cf383378bbd8a2f378b6

**Comment by ls-zhu** [comment]
close and re-open for trigger Travis again

**Comment by ls-zhu** [comment]
reopen for trigger Travis again

**Comment by krig** [comment]
I ran the CI tests successfully in a branch on my github user, so merging this now.

**Comment by ls-zhu** [comment]
Thanks!

**Pull request merged**